File: sort_efficient/sort_efficient__quick.cs
// 実行時間 410ms
using System.Linq;
using static System.Console;

internal class Program
{
    static int o = 0;

    static void Main()
    {
        int n = int.Parse( ReadLine() );
        int[] a = ReadLine().Split().Select( int.Parse ).ToArray();

        QuickSort( a, 0, n );
        WriteLine( string.Join( " ", a ) );
        WriteLine( o );
        return;
    }

    static void QuickSort( int[] a, int l, int r )
    {
        if( r <= l + 1 ) {
            return;
        }

        int c = l;
        int p = a[r - 1];
        int t = 0;

        for( int i = l; i <= r - 2; i++ ) {
            if( a[i] < p ) {
                t = a[c];
                a[c] = a[i];
                a[i] = t;
                c++;
                o++;
            }
        }
        t = a[c];
        a[c] = a[r - 1];
        a[r - 1] = t;
        QuickSort( a, l, c );
        QuickSort( a, c + 1, r );
        return;
    }
}

File: sort_efficient/sort_efficient__quick.py
# 実行時間 1680ms
def quickSort(a, l, r):
    if r <= l + 1:
        return

    global o
    c = l
    p = a[r - 1]

    for i in range(l, r - 1):
        if a[i] < p:
            a[c], a[i] = a[i], a[c]
            c += 1
            o += 1
    a[c], a[r - 1] = a[r - 1], a[c]
    quickSort(a, l, c)
    quickSort(a, c + 1, r)

n = int(input())
o = 0
a = list(map(int, input().split()))

quickSort(a, 0, n)
print(*a)
print(o)
